docs/design-documents/features/connectivity/TLSSocket - Design document.md

```diff
@@ -1,26 +1,31 @@
 # TLSSocket - Design document
 ## Overview and Background
 
-This document describes design of TLSSocket class that provides simple interface for Mbed OS
-user to create TLS connections over TCP socket.
+This document describes design of TLSSocket class that provides simple interface for Mbed OS user to create TLS connections over TCP socket.
+
+Mbed OS provides 2 different implementations for TLS socket; Mbed TLS based (default) or offloaded TLS socket when used network stack (e.g. cellular modem target driver) supports it.
 
 This class greatly simplifies the usage of TLS but limits itself to only one use case.
-This design limitation is accepted as other users can continue using Mbed TLS API
-directly. TLSSocket also exposes internal Mbed TLS structures allowing use of Mbed TLS API to configure the underlying library.
+This design limitation is accepted as other users can continue using Mbed TLS API directly. Mbed TLS based TLSSocket also exposes internal Mbed TLS structures allowing use of  Mbed TLS API to configure the underlying library.
 
 High level goal is to demonstrate Mbed OS users that secure connections are not hard to do.
 
 ## Design Limitations
 
-Following design limitations exist in the current design of TLSSocket
+Following design limitations exist in the current design of TLSSocket:
 
+Mbed TLS based TLSSocket:
 * `TLSSocket::connect()` is always blocking
 * Can only use server and client certificates through `set_root_ca_cert()` and `set_client_cert_key()` methods. For other use cases, internal Mbed TLS structures are exposed.
 * No PSK mode
 
+Offloaded vs. Mbed TLS based TLSSocket:
+* For offloaded TLS socket `set_root_ca_cert()` and `set_client_cert_key()` must be called after `TLSSocket::open()` and before `TLSSocket::connect()`.
+* Offloaded TLS socket API does not support all Mbed TLS based TLSSocket methods, but commot (i.e. `open()`, `connect()`, `close()` and setting certficates) use the same API.
+
 # System Architecture and High-Level Design
 
-Secure socket consist of two classes:
+Mbed TLS based secure socket consist of two classes:
 
 * `TLSocketWrapper` <br>
     Which handles initialization of TLS library and does TLS handsake. Takes any Socket as a
```

Comment:
I know this has been here before, but it is actually not true: We have a `DTLSSocketWrapper` for datagram sockets. Perhaps we could add an explanation here?

Reply:
Also `DTLSSocket` needs to be added. But I'll create a new internal task add those, so let's leave those out from this PR.